include/envoy/server/resource_monitor.h

```diff
@@ -18,39 +21,106 @@ struct ResourceUsage {
+  virtual ~ResourceUpdateCallbacks() = default;
 
   /**
-   * Notifies caller of updated resource usage.
+   * Called when the request for updated resource usage succeeds.
+   * @param usage the updated resource usage
    */
-  class Callbacks {
-  public:
-    virtual ~Callbacks() = default;
-
-    /**
-     * Called when the request for updated resource usage succeeds.
-     * @param usage the updated resource usage
-     */
-    virtual void onSuccess(const ResourceUsage& usage) PURE;
-
-    /**
-     * Called when the request for updated resource usage fails.
-     * @param error the exception caught when trying to get updated resource usage
-     */
-    virtual void onFailure(const EnvoyException& error) PURE;
-  };
+  virtual void onSuccess(const ResourceUsage& usage) PURE;
+
+  /**
+   * Called when the request for updated resource usage fails.
+   * @param error the exception caught when trying to get updated resource usage
+   */
+  virtual void onFailure(const EnvoyException& error) PURE;
+};
+
+class ReactiveResourceUpdateCallbacks {
+public:
+  virtual ~ReactiveResourceUpdateCallbacks() = default;
+
+  /**
+   * Called when the request for updated resource usage succeeds.
+   * @param usage the updated resource usage
+   */
+  virtual void onSuccess(const uint64_t usage) PURE;
+
+  /**
+   * Called when the request for updated resource usage fails.
+   * todo may need propagate error
+   */
+  virtual void onFailure() PURE;
+
+  virtual Event::Dispatcher& dispatcher() PURE;
+};
+
+class ResourceMonitor {
+public:
+  virtual ~ResourceMonitor() = default;
 
   /**
    * Recalculate resource usage.
    * This must be non-blocking so if RPCs need to be made they should be
    * done asynchronously and invoke the callback when finished.
    */
-  virtual void updateResourceUsage(Callbacks& callbacks) PURE;
+  virtual void updateResourceUsage(ResourceUpdateCallbacks& callbacks) PURE;
+};
+
+class ReactiveResourceMonitor {
+public:
+  ReactiveResourceMonitor() = default;
+  virtual ~ReactiveResourceMonitor() = default;
+  virtual bool tryAllocateResource(uint64_t increment) PURE;
+  virtual bool tryDeallocateResource(uint64_t decrement) PURE;
+  virtual uint64_t currentResourceUsage() const PURE;
+  virtual uint64_t maxResourceUsage() const PURE;
 };
 
 using ResourceMonitorPtr = std::unique_ptr<ResourceMonitor>;
 
+using ReactiveResourceMonitorPtr = std::unique_ptr<ReactiveResourceMonitor>;
+
+// Example of reactive resource monitor. To be removed.
+class ActiveConnectionsResourceMonitor : public ReactiveResourceMonitor {
+public:
+  ActiveConnectionsResourceMonitor(uint64_t max_active_conns)
+      : max_(max_active_conns), current_(0){};
+
+  bool tryAllocateResource(uint64_t increment) {
+    auto curr_capacity = current_.load();
+    if (std::numeric_limits<uint64_t>::max() - increment < curr_capacity) {
+      // uint64_t overflow, cannot allocate resource.
+      return false;
+    } else {
+      current_ += increment;
```

Comment:
I think that it is possible for multiple threads to hit this increment concurrently after narrowly passing the checks above to verify that resource is available.  Say that current_ is `max_ - 1` as N threads attempt to increment by 1, all N of those threads may execute the current_.load() before the first thread gets to the increment, so at the end of these tryAllocateResource calls the value of `current_` could be any value in `[max_, max_ + N - 1]`.

The question how strict we want to be about this limit.  Possible way to be more strict:
```
uint64_t new_val = (current_ += increment);
if (new_val >= max_) {
  current_ -= increment;
  return false;
}
return true;
```

Reply:
Double guard would also protect against potential race, though this version has more code:
```
    if (current_.load() + inrement > max) {
      return false;
    } else {
      //Guard against raice condition.
      if (current_.load() + inrement <= max) {
        current_ += increment;
        return true;
      }
      return false;
    }
```